dev/AppNotifications/ShellLocalization.cpp

```diff
@@ -0,0 +1,272 @@
+
+    std::transform(lowercaseFileExtension.begin(), lowercaseFileExtension.end(), lowercaseFileExtension.begin(),
+        [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
+
+    const bool isFileExtensionSupported =
+        lowercaseFileExtension == ".ico" || lowercaseFileExtension == ".bmp" || lowercaseFileExtension == ".jpg" || lowercaseFileExtension == ".png";
+
+    THROW_HR_IF(E_UNEXPECTED, isFileExtensionSupported);
+    return S_OK;
+}
+
+inline std::wstring RetrieveLocalFolderPath()
+{
+    wil::unique_cotaskmem_string localAppDataPath;
+    THROW_IF_FAILED(SHGetKnownFolderPath(FOLDERID_LocalAppData, 0 /* flags */, nullptr /* access token handle */, &localAppDataPath));
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK
+    std::path localFolderPath{ std::wstring(localAppDataPath.get()) +
+        Microsoft::Windows::AppNotifications::Helpers::c_localMicrosoftFolder +
+        Microsoft::Windows::AppNotifications::Helpers::c_localWindowsAppSDKFolder };
+
+    if (!std::exists(localFolderPath))
+    {
+        std::create_directory(localFolderPath);
+    }
+
+    return std::wstring(localFolderPath.c_str());
+}
+
+inline wil::unique_hicon RetrieveIconFromProcess()
+{
+    std::wstring processPath{};
+    THROW_IF_FAILED(wil::GetModuleFileNameExW(GetCurrentProcess(), nullptr, processPath));
+
+    // Extract the small icon from the first .ico, if failed extract the large icon.
+    // Small icon is good enough for an App Notification icon since higher quality doesn't really impact in a substantial way.
+    wil::unique_hicon hIcon{};
+    if (!ExtractIconExW(processPath.c_str(), 0 /* index */, nullptr /* Large icon */, &hIcon, 1))
+    {
+        THROW_HR_IF(E_FAIL, ExtractIconExW(processPath.c_str(), 0, &hIcon, nullptr /* Small Icon */, 1) == 0);
+    }
+
+    return hIcon;
+}
+
+HRESULT Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveAssetsFromProcess(_Out_ Microsoft::Windows::AppNotifications::ShellLocalization::AppNotificationAssets& assets) noexcept try
+{
+    wil::unique_hicon hIcon{ RetrieveIconFromProcess() };
+
+    std::wstring notificationAppId{ Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK\{AppGUID}.png
+    std::wstring iconFilePath{ RetrieveLocalFolderPath().c_str() + c_backSlash + notificationAppId + c_pngExtension };
+    WriteHIconToPngFile(hIcon, iconFilePath.c_str());
+
+    assets.displayName = Microsoft::Windows::AppNotifications::Helpers::GetDisplayNameBasedOnProcessName();
+    assets.iconFilePath = iconFilePath;
+
+    // TODO: In case we are caching icons from uninstalled apps who didn't call UnregisterAll,
+    // we should make an effort to clean them up in an opportunistic way.
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+// Do nothing. This is just a placeholder while the UDK is ingested with the proper API.
```

Comment:
So this will be updated when the UDK's updated?

Move the comment inside the function and change it to `//TODO: This is a placeholder until the UDK is ingested with the proper API`

Even better if there's an ADO item for that work make it `//TODO:12345 Use UDK API to retrieve icon` or whatever the ADO item # and title are

Reply:
> So this will be updated when the UDK's updated?

Yes. But it's not blocking the PR.

The change has been committed, now we are waiting for a fresh UDK that contains the API.